packages/backend/server/src/plugins/copilot/context/session.ts

```diff
@@ -0,0 +1,236 @@
+import { OneMB } from '../../../core/quota/constant';
+import {
+  ContextConfig,
+  ContextFile,
+  ContextFileStatus,
+  DocChunkSimilarity,
+  Embedding,
+  EmbeddingClient,
+  FileChunkSimilarity,
+} from './types';
+
+export class ContextSession implements AsyncDisposable {
+  constructor(
+    private readonly client: EmbeddingClient,
+    private readonly contextId: string,
+    private readonly config: ContextConfig,
+    private readonly db: PrismaClient
+  ) {}
+
+  get id() {
+    return this.contextId;
+  }
+
+  get workspaceId() {
+    return this.config.workspaceId;
+  }
+
+  async listDocs() {
+    return [...this.config.docs];
+  }
+
+  async listFiles() {
+    return this.config.files.map(f => ({ ...f }));
+  }
+
+  private processEmbeddings(fileId: string, embeddings: Embedding[]) {
+    const groups = embeddings.map(e => [
+      randomUUID(),
+      this.contextId,
+      fileId,
+      e.index,
+      e.content,
+      Prisma.raw(`'[${e.embedding.join(',')}]'`),
+      new Date(),
+    ]);
+    return Prisma.join(groups.map(row => Prisma.sql`(${Prisma.join(row)})`));
+  }
+
+  private async insertEmbeddings(
+    name: string,
+    blobId: string,
+    embeddings: Embedding[]
+  ) {
+    const fileId = nanoid();
+    await this.saveFileRecord(fileId, file => ({
+      ...file,
+      blobId,
+      chunk_size: embeddings.length,
+      name,
+    }));
+
+    const values = this.processEmbeddings(fileId, embeddings);
+    return this.db.$transaction(async tx => {
+      await tx.$executeRaw`
+        INSERT INTO "ai_context_embeddings"
+        ("id", "context_id", "file_id", "chunk", "content", "embedding", "updated_at") VALUES ${values}
+        ON CONFLICT (context_id, file_id, chunk) DO UPDATE SET
+        content = EXCLUDED.content, embedding = EXCLUDED.embedding, updated_at = excluded.updated_at;
+      `;
+      await this.saveFileRecord(
+        fileId,
+        file => ({
+          ...(file as ContextFile),
+          status: ContextFileStatus.finished,
+        }),
+        tx
+      );
+      return fileId;
+    });
+  }
+
+  private readStream(
+    readable: Readable,
+    maxSize = 50 * OneMB
+  ): Promise<Buffer<ArrayBuffer>> {
+    return new Promise<Buffer<ArrayBuffer>>((resolve, reject) => {
+      const chunks: Uint8Array[] = [];
+      let totalSize = 0;
+
+      readable.on('data', chunk => {
+        totalSize += chunk.length;
+        if (totalSize > maxSize) {
+          reject(new BlobQuotaExceeded());
+          readable.destroy(new BlobQuotaExceeded());
+          return;
+        }
+        chunks.push(chunk);
+      });
+
+      readable.on('end', () => {
+        resolve(Buffer.concat(chunks, totalSize));
+      });
+
+      readable.on('error', err => {
+        reject(err);
+      });
+    });
+  }
+
+  async addDocRecord(docId: string) {
+    if (!this.config.docs.includes(docId)) {
+      this.config.docs.push(docId);
+      await this.save();
+    }
+    return this.config.docs.length;
```

Comment:
Better to return `docId` and `createTime` for this API. No need for array length in front-end.

Reply:
frontend needs to use the result of the list as the standard